pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,546 @@
+	orderedSvcExports := make([]*mcsapitypes.MCSAPIServiceSpec, 0, len(svcExportByCluster))
+	for _, mcsAPISvcSpec := range svcExportByCluster {
+		orderedSvcExports = append(orderedSvcExports, mcsAPISvcSpec)
+	}
+	slices.SortFunc(orderedSvcExports, func(a, b *mcsapitypes.MCSAPIServiceSpec) int {
+		if a.ExportCreationTimestamp.Equal(&b.ExportCreationTimestamp) {
+			return strings.Compare(a.Cluster, b.Cluster)
+		}
+		if a.ExportCreationTimestamp.Before(&b.ExportCreationTimestamp) {
+			return -1
+		}
+		return 1
+	})
+	return orderedSvcExports
+}
+
+func isSimilarPort(a, b mcsapiv1alpha1.ServicePort) bool {
+	return a.Port == b.Port && a.Protocol == b.Protocol
+}
+
+func findSimilarPort(port mcsapiv1alpha1.ServicePort, ports []portMerge) int {
+	for i, currPort := range ports {
+		if isSimilarPort(port, currPort.ServicePort) {
+			return i
+		}
+	}
+	return -1
+}
+
+// checkDuplicatedPortNameConflict check if port has a duplicated port name in portsByName.
+func checkDuplicatedPortNameConflict(port portMerge, portsByName map[string]portMerge) (foundPortByName bool, hasConflict bool, conflictMsg string) {
+	portByName, foundPortByName := portsByName[port.Name]
+	if foundPortByName && !isSimilarPort(port.ServicePort, portByName.ServicePort) {
+		hasConflict = true
+		return foundPortByName, hasConflict, fmt.Sprintf(
+			"Duplicated port name \"%s\". Using port definition \"%v\" from oldest service export in cluster \"%s\".",
+			port.Name, portByName, portByName.cluster,
+		)
+	}
+
+	return foundPortByName, hasConflict, ""
+}
+
+func checkPortConflict(port, olderPort portMerge) string {
+	if port.Name != olderPort.Name {
+		return fmt.Sprintf(
+			"Conflicting port name on \"%v\". Using the port name \"%s\" from oldest service export in cluster \"%s\".",
+			port, olderPort.Name, olderPort.cluster,
+		)
+	}
+	if ptr.Deref(port.AppProtocol, "") != ptr.Deref(olderPort.AppProtocol, "") {
+		return fmt.Sprintf(
+			"Conflicting appProtocol on \"%v\". Using the appProtocol \"%s\" from oldest service export in cluster \"%s\".",
+			port, ptr.Deref(olderPort.AppProtocol, ""), olderPort.cluster,
+		)
+	}
+	return ""
+}
+
+// mergePorts merge all the ports into a map while doing conflict resolution
+// with the oldest CreationTimestamp. It also return if it detects any conflict
+func mergePorts(svcExportByCluster operator.ServiceExportsByCluster) ([]portMerge, string) {
+	conflict := ""
+	orderedSvcExports := orderSvcExportByPriority(svcExportByCluster)
+	ports := []portMerge{}
+	portsByName := map[string]portMerge{}
+	for _, svcExport := range orderedSvcExports {
+		for _, port := range svcExport.Ports {
+			portMergeValue := portMerge{
+				ServicePort:             port,
+				cluster:                 svcExport.Cluster,
+				ExportCreationTimestamp: svcExport.ExportCreationTimestamp,
+			}
+
+			foundPortByName := true
+			if port.Name != "" {
```

Comment:
Any reason why this check is skipped if the port name is empty? I'd intuitively treat it the same of any other name, which should also allow to simplify the logic a bit. Related, I guess there should be an extra check to assert that if the empty port is present, then it is the only one, otherwise the service creation would be rejected.

Reply:
So this is a bit confusing because there is two logic that checks the port name. This part of the logic take two port that has the same name and tries to check if it is actually the same port (aka same port/proto). This means that in the case of an unnamed port we don't have to do this because we can have multiple unnamed port that are different which is a valid thing afaik.

And the other part below that case does the opposite, it takes two similar port and check if every properties match including port name (here would be the check https://github.com/MrFreezeex/cilium/blob/7d8f382952896570fa8529655688876712237808/pkg/clustermesh/mcsapi/serviceimport_controller.go#L196). And this AFAIU would trigger the conflict on the case you just presented (same port one unnamed and one named).

Reply:
> This means that in the case of an unnamed port we don't have to do this because we can have multiple unnamed port that are different which is a valid thing afaik.

The Kubernetes API Server actually rejects a service with multiple unnamed ports:

```bash
$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  - port: 90
    protocol: TCP
    targetPort: 90
  selector:
    app: foo
EOF

The Service "foo" is invalid: 
* spec.ports[0].name: Required value
* spec.ports[1].name: Required value
```

And the same also applies if there are multiple ports, and any is unnamed.

```bash
$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  - name: bar
    port: 90
    protocol: TCP
    targetPort: 90
  selector:
    app: foo
EOF

The Service "foo" is invalid: spec.ports[0].name: Required value
```

Reply:
Ahhh TIL indeed thanks, I will update it thanks!